datameta/api/metadatasets.py

```diff
@@ -136,6 +166,136 @@ def delete_metadatasets(request: Request) -> HTTPNoContent:
+
+    # Render records according to MetaDatum constraints.
+    record = render_record_values(metadata, record)
+
+    # construct new MetaDataSet:
+    mdata_set = MetaDataSet(
+        site_id = siteid.generate(request, MetaDataSet),
+        user_id = auth_user.id,
+        submission_id = None
+    )
+
+    db.add(mdata_set)
+
+    mset_repl_evt = MsetReplacementEvent(
+        user_id = auth_user.id,
+        datetime = datetime.utcnow(),
+        label = replaces_label,
+        new_metadataset_id = mdata_set.id
+    )
+    db.add(mset_repl_evt)
+
+    msets = [
+        (mset_id, resource_by_id(db, MetaDataSet, mset_id))
+        for mset_id in replaces
+    ]
+
+    missing_msets = [("Invalid metadataset id.", mset_id) for mset_id, target_mset in msets if target_mset is None]
+
+    if missing_msets:
+        messages, entities = zip(*missing_msets)
+        raise errors.get_validation_error(messages=messages, entities=entities)
+
+    already_replaced = [
+        (f"Metadataset already replaced by {get_identifier(target_mset.replaced_via_event.new_metadataset)}", mset_id)
+        for mset_id, target_mset in msets
+        if target_mset.replaced_via_event_id is not None
+    ]
+
+    if already_replaced:
+        messages, entities = zip(*already_replaced)
+        raise errors.get_validation_error(messages=messages, entities=entities)
+
+    for _, target_mset in msets:
+        target_mset.replaced_via_event_id = mset_repl_evt.id
+
+    # Add NULL values for service metadata
+    service_metadata = get_service_metadata(db)
+    for s_mdatum in service_metadata.values():
+        mdatum_rec = MetaDatumRecord(
+                metadatum_id     = s_mdatum.id,
+                metadataset_id   = mdata_set.id,
+                file_id          = None,
+                value            = None
+                )
+        db.add(mdatum_rec)
+
+    # Add the non-service metadata as specified in the request body
+    for name, value in record.items():
+        mdatum_rec = MetaDatumRecord(
+            metadatum_id     = metadata[name].id,
+            metadataset_id   = mdata_set.id,
+            file_id          = None,
+            value            = value
+        )
+        db.add(mdatum_rec)
+
+    # Collect files, drop duplicates
+    file_ids = set(request.openapi_validated.body['fileIds'])
+    db_files = { file_id : resource.resource_query_by_id(db, File, file_id).options(joinedload(File.metadatumrecord)).one_or_none() for file_id in file_ids }
+
+    # Validate submission access to the specified files
+    validation.validate_submission_access(db, db_files, {}, auth_user)
+
+    # Validate the provided records
+    validate_metadataset_record(service_metadata, record, return_err_message=False, rendered=False)
+
+    # Validate the associations between files and records
+    fnames, ref_fnames, val_errors = validation.validate_submission_association(db_files, { mdata_set.site_id : mdata_set }, ignore_submitted_metadatasets=True)
```

Comment:
We probably don't need to set this, since the freshly generated metadataset won't be submitted yet
```suggestion
    fnames, ref_fnames, val_errors = validation.validate_submission_association(db_files, { mdata_set.site_id : mdata_set })
```